app/assets/javascripts/discourse/app/mixins/composer-upload-uppy.js

```diff
@@ -0,0 +1,468 @@
+      this._uppyInstance.cancelAll();
+      this._uppyInstance = null;
+    }
+  },
+
+  _bindUploadTarget() {
+    this.placeholders = {};
+    this._preProcessorStatus = {};
+    this.fileInputEl = document.getElementById("file-uploader");
+    const isPrivateMessage = this.get("composer.privateMessage");
+
+    this.appEvents.on("composer:add-files", this._addFiles.bind(this));
+
+    this._unbindUploadTarget();
+    this._bindFileInputChangeListener();
+    this._bindPasteListener();
+    this._bindMobileUploadButton();
+
+    this._uppyInstance = new Uppy({
+      id: "composer-uppy",
+      autoProceed: true,
+
+      // need to use upload_type because uppy overrides type with the
+      // actual file type
+      meta: deepMerge({ upload_type: "composer" }, this.data || {}),
+
+      onBeforeFileAdded: (currentFile) => {
+        const validationOpts = {
+          user: this.currentUser,
+          siteSettings: this.siteSettings,
+          isPrivateMessage,
+          allowStaffToUploadAnyFileInPm: this.siteSettings
+            .allow_staff_to_upload_any_file_in_pm,
+        };
+
+        const isUploading = validateUploadedFile(currentFile, validationOpts);
+
+        run(() => {
+          this.setProperties({
+            uploadProgress: 0,
+            isUploading,
+            isCancellable: isUploading,
+          });
+        });
+
+        if (!isUploading) {
+          this.appEvents.trigger("composer:uploads-aborted");
+        }
+        return isUploading;
+      },
+
+      onBeforeUpload: (files) => {
+        const fileCount = Object.keys(files).length;
+        const maxFiles = this.siteSettings.simultaneous_uploads;
+
+        // Limit the number of simultaneous uploads
+        if (maxFiles > 0 && fileCount > maxFiles) {
+          bootbox.alert(
+            I18n.t("post.errors.too_many_dragged_and_dropped_files", {
+              count: maxFiles,
+            })
+          );
+          this.appEvents.trigger("composer:uploads-aborted");
+          this._reset();
+          return false;
+        }
+      },
+    });
+
+    this._uppyInstance.use(DropTarget, { target: this.element });
+    this._uppyInstance.use(UppyChecksum, { capabilities: this.capabilities });
+
+    // TODO (martin) Need a more automatic way to do this for preprocessor
+    // plugins like UppyChecksum and UppyMediaOptimization so people don't
+    // have to remember to do this, also want to wrap this.uppy.emit in those
+    // classes so people don't have to remember to pass through the plugin class
+    // name for the preprocess-X events.
+    this._trackPreProcessorStatus(UppyChecksum);
+
+    // TODO (martin) support for direct S3 uploads will come later, for now
+    // we just want the regular /uploads.json endpoint to work well
+    this._useXHRUploads();
+
+    // TODO (martin) develop upload handler guidance and an API to use; will
+    // likely be using uppy plugins for this
+    this._uppyInstance.on("file-added", (file) => {
```

Comment:
The 3 above todos I will address in another PR...perfect is the enemy of the good, I need to get this in front of a real world environment.